crates/sui-framework/sources/display.move

```diff
@@ -0,0 +1,184 @@
+    /// ```
+    /// // Example of a display object
+    /// Display<0x...::capy::Capy> {
+    ///  fields:
+    ///    <name, "Capy {{ genes }}">
+    ///    <link, "https://capy.art/capy/{{ id }}">
+    ///    <image, "https://api.capy.art/capy/{{ id }}/svg">
+    ///    <description, "Lovely Capy, one of many">
+    /// }
+    /// ```
+    ///
+    /// Uses only String type due to external-facing nature of the object,
+    /// the property names have a priority over their types.
+    struct Display<phantom T: key> has key {
+        id: UID,
+        /// Contains fields for display. Currently supported
+        /// fields are: name, link, image and description.
+        fields: VecMap<String, String>,
+    }
+
+    /// Event: emitted when a new Display object has been created for type T.
+    /// Type signature of the event corresponds to the type while id serves for
+    /// the discovery.
+    struct DisplayCreated<phantom T: key> has copy, drop {
+        id: ID
+    }
+
+    // ============ Initializer Functions ===========
+
+    /// Since the only way to own a Display is before it has been published,
+    /// we don't need to perform an authorization check.
+    ///
+    /// Also, the only place it can be used is the function where the Display
+    /// object was created; hence values and names are likely to be hardcoded and
+    /// vector<u8> is the best type for that purpose.
+    public fun set_owned<T: key>(d: Display<T>, name: vector<u8>, value: vector<u8>): Display<T> {
+        set_internal(&mut d, utf8(name), utf8(value));
+        d
+    }
+
+    /// Create an empty Display object. It can either be shared empty of filled
+    /// with data right away via cheaper `set_owned` method.
+    public fun empty<T: key>(pub: &Publisher, ctx: &mut TxContext): Display<T> {
+        assert!(is_package<T>(pub), ENotOwner);
+
+        let uid = object::new(ctx);
+
+        event::emit(DisplayCreated<T> {
+            id: object::uid_to_inner(&uid)
+        });
+
+        Display {
+            id: uid,
+            fields: vec_map::empty()
+        }
+    }
+
+    /// Share an object after the initialization is complete.
+    public fun share<T: key>(d: Display<T>) {
+        transfer::share_object(d);
+    }
+
+    // ========== Entry functions ==========
+
+    /// Create a new empty Display<T> object and share it.
+    entry public fun create_and_share<T: key>(pub: &Publisher, ctx: &mut TxContext) {
+        share(empty<T>(pub, ctx))
+    }
+
+    /// Create a new Display<T> object with a set of fields.
+    entry public fun create_with_fields<T: key>(
+        pub: &Publisher, fields: vector<String>, values: vector<String>, ctx: &mut TxContext
+    ) {
+        let len = vector::length(&fields);
+        assert!(len == vector::length(&values), EVecLengthMismatch);
+
+        let display = empty<T>(pub, ctx);
+        while (len > 0) {
+            set_internal(&mut display, *vector::borrow(&fields, len), *vector::borrow(&values, len));
```

Comment:
Won't `len` be out of bounds on the first iteration of the loop?

I typically do `while (i < len)` to avoid this (though of course we should/will have looping primitives that will avoid all of this silliness).